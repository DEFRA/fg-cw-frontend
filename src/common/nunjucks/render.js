import camelCase from "lodash/camelCase.js";
import { environment } from "./nunjucks.js";

/**
 * @param {string} name
 * @param {object} params
 * @param {string} [block]
 */
export const render = (name, params, block) => {
  const macro = {
    path: `${name}/macro.njk`,
    name: camelCase(name.split("-")),
    params: JSON.stringify(params, null, 2),
  };

  const callComponent = block
    ? `{%- call ${macro.name}(${macro.params}) -%}${block}{%- endcall -%}`
    : `{{- ${macro.name}(${macro.params}) -}}`;

  return environment.renderString(
    `{%- from "${macro.path}" import ${macro.name} -%}
    ${callComponent}
    `,
    {},
  );
};
