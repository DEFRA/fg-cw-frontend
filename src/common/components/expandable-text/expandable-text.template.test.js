import { describe, expect, test } from "vitest";
import { render } from "../../nunjucks/render.js";

describe("expandable-text template", () => {
  test("renders truncated text with toggle when text is long", () => {
    const longText =
      "This is a very long piece of text that should definitely be truncated because it exceeds the default truncate length of 100 characters and should show a toggle button.";

    const component = render("expandable-text", {
      text: longText,
      href: "/some-link",
      ref: "test-ref-123",
      isSelected: false,
    });

    expect(component).toMatchSnapshot();
  });

  test("renders full text without toggle when text is short", () => {
    const shortText = "This is short text.";

    const component = render("expandable-text", {
      text: shortText,
      href: "/some-link",
      ref: "test-ref-456",
      isSelected: false,
    });

    expect(component).toMatchSnapshot();
  });

  test("renders full text when isSelected is true regardless of length", () => {
    const longText =
      "This is a very long piece of text that should definitely be truncated because it exceeds the default truncate length of 100 characters but won't because isSelected is true.";

    const component = render("expandable-text", {
      text: longText,
      href: "/some-link",
      ref: "test-ref-789",
      isSelected: true,
    });

    expect(component).toMatchSnapshot();
  });

  test("renders with custom linkText and truncateLength", () => {
    const longText =
      "This text will be truncated at 50 characters because we set a custom truncate length parameter.";

    const component = render("expandable-text", {
      text: longText,
      href: "/custom-link",
      ref: "test-ref-custom",
      isSelected: false,
      linkText: "show more details",
      truncateLength: 50,
    });

    expect(component).toMatchSnapshot();
  });

  test("renders without toggle when href is missing", () => {
    const longText =
      "This is a very long piece of text that should be truncated but won't have a toggle button because href is missing from the parameters.";

    const component = render("expandable-text", {
      text: longText,
      ref: "test-ref-no-href",
      isSelected: false,
    });

    expect(component).toMatchSnapshot();
  });

  test("renders with correct data attributes", () => {
    const longText =
      "This text tests that the correct data attributes are set on the container div element.";

    const component = render("expandable-text", {
      text: longText,
      href: "/data-test",
      ref: "test-ref-data",
      isSelected: false,
      linkText: "read more",
      truncateLength: 75,
    });

    expect(component).toMatchSnapshot();
  });
});
