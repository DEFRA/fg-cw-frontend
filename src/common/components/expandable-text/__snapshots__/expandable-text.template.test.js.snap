// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`expandable-text template > renders full text when isSelected is true regardless of length 1`] = `
"
<div
  data-module="expandable-text"
  data-truncate-length="100"
  data-expand-text="read full text"
>
    <span class="expandable-text__content">This is a very long piece of text that should definitely be truncated because it exceeds the default truncate length of 100 characters but won&#39;t because isSelected is true.</span>
</div>
"
`;

exports[`expandable-text template > renders full text without toggle when text is short 1`] = `
"
<div
  data-module="expandable-text"
  data-truncate-length="100"
  data-expand-text="read full text"
>
    <span class="expandable-text__content">This is short text.</span>
</div>
"
`;

exports[`expandable-text template > renders truncated text with toggle when text is long 1`] = `
"
<div
  data-module="expandable-text"
  data-truncate-length="100"
  data-expand-text="read full text"
>
    <span
      class="expandable-text__content"
      id="expandable-text-content-test-ref-123"
      >This is a very long piece of text that should definitely be truncated because it exceeds the default...</span
    >
    <span
      class="expandable-text__full-content"
      id="expandable-text-full-test-ref-123"
      hidden
      >This is a very long piece of text that should definitely be truncated because it exceeds the default truncate length of 100 characters and should show a toggle button.</span
    >
      <a
        href="/some-link"
        class="expandable-text__toggle"
        role="button"
        aria-expanded="false"
        aria-controls="expandable-text-content-test-ref-123"
        aria-describedby="expandable-text-full-test-ref-123"
      >
        read full text
      </a>
</div>
"
`;

exports[`expandable-text template > renders with correct data attributes 1`] = `
"
<div
  data-module="expandable-text"
  data-truncate-length="75"
  data-expand-text="read more"
>
    <span
      class="expandable-text__content"
      id="expandable-text-content-test-ref-data"
      >This text tests that the correct data attributes are set on the container d...</span
    >
    <span
      class="expandable-text__full-content"
      id="expandable-text-full-test-ref-data"
      hidden
      >This text tests that the correct data attributes are set on the container div element.</span
    >
      <a
        href="/data-test"
        class="expandable-text__toggle"
        role="button"
        aria-expanded="false"
        aria-controls="expandable-text-content-test-ref-data"
        aria-describedby="expandable-text-full-test-ref-data"
      >
        read more
      </a>
</div>
"
`;

exports[`expandable-text template > renders with custom linkText and truncateLength 1`] = `
"
<div
  data-module="expandable-text"
  data-truncate-length="50"
  data-expand-text="show more details"
>
    <span
      class="expandable-text__content"
      id="expandable-text-content-test-ref-custom"
      >This text will be truncated at 50 characters becau...</span
    >
    <span
      class="expandable-text__full-content"
      id="expandable-text-full-test-ref-custom"
      hidden
      >This text will be truncated at 50 characters because we set a custom truncate length parameter.</span
    >
      <a
        href="/custom-link"
        class="expandable-text__toggle"
        role="button"
        aria-expanded="false"
        aria-controls="expandable-text-content-test-ref-custom"
        aria-describedby="expandable-text-full-test-ref-custom"
      >
        show more details
      </a>
</div>
"
`;

exports[`expandable-text template > renders without toggle when href is missing 1`] = `
"
<div
  data-module="expandable-text"
  data-truncate-length="100"
  data-expand-text="read full text"
>
    <span
      class="expandable-text__content"
      id="expandable-text-content-test-ref-no-href"
      >This is a very long piece of text that should be truncated but won&#39;t have a toggle button because hr...</span
    >
    <span
      class="expandable-text__full-content"
      id="expandable-text-full-test-ref-no-href"
      hidden
      >This is a very long piece of text that should be truncated but won&#39;t have a toggle button because href is missing from the parameters.</span
    >
</div>
"
`;
